pkg/handler/metatile.go

```diff
@@ -86,134 +92,245 @@ func MetatileHandler(
-			// the net/http package exposes.
-			lastModifiedFormatted := lastMod.UTC().Format(http.TimeFormat)
-			headers.Set("Last-Modified", lastModifiedFormatted)
-			reqState.StorageMetadata.HasLastModified = true
-		}
-		if etag := storageResp.ETag; etag != nil {
-			headers.Set("ETag", *etag)
-			reqState.StorageMetadata.HasEtag = true
+
+		if metatileResponseData.ResponseState == state.ResponseState_NotFound {
+			http.NotFound(rw, req)
+			reqState.ResponseState = state.ResponseState_NotFound
+			return
+		} else if metatileResponseData.ResponseState == state.ResponseState_NotModified {
+			rw.WriteHeader(http.StatusNotModified)
+			reqState.ResponseState = state.ResponseState_NotModified
+			return
 		}
 
-		metatileReaderFindStart := time.Now()
-		reader, formatSize, err := tile.NewMetatileReader(offset, bytes.NewReader(storageBytes), bodySize)
-		reqState.Duration.MetatileFind = time.Since(metatileReaderFindStart)
+		responseData, err := extractVectorTileFromMetatile(reqState, bufferManager, parseResult, metatileResponseData)
 		if err != nil {
-			logger.Error(log.LogCategory_MetatileError, "Failed to read metatile: %#v", err)
 			http.Error(rw, err.Error(), http.StatusInternalServerError)
-			reqState.IsZipError = true
 			reqState.ResponseState = state.ResponseState_Error
 			return
 		}
-		// make sure to close zip file reader
-		defer reader.Close()
-		reqState.ResponseSize = int(formatSize)
-		headers.Set("Content-Length", fmt.Sprintf("%d", formatSize))
-
-		rw.WriteHeader(http.StatusOK)
-		reqState.ResponseState = state.ResponseState_Success
-		respWriteStart := time.Now()
-		_, err = io.Copy(rw, reader)
-		reqState.Duration.RespWrite = time.Since(respWriteStart)
+
+		// Copy some of the metatile response data over to the vector tile response data so that it is properly cached
+		responseData.ETag = metatileResponseData.ETag
+		responseData.LastModified = metatileResponseData.LastModified
+
+		err = writeVectorTileResponse(reqState, rw, responseData)
 		if err != nil {
+			// TODO Context cancellation might happen here?
 			logger.Error(log.LogCategory_ResponseError, "Failed to write response body: %#v", err)
-			reqState.IsResponseWriteError = true
+			// Still want to set the cache in this case
+		}
+
+		// Cache the response
+		// TODO Do this in a goroutine so the handler can exit faster?
+		cacheSetStart := time.Now()
+		timeoutCtx, cancel = context.WithTimeout(req.Context(), cacheTimeout)
+		err = tileCache.SetTile(timeoutCtx, parseResult, responseData)
+		cancel()
+		reqState.Duration.CacheSet = time.Since(cacheSetStart)
+		if err != nil {
+			logger.Error(log.LogCategory_ResponseError, "Failed to set cache: %#v", err)
 		}
 	})
 }
 
+func fetchMetatile(reqState *state.RequestState, stg storage.Storage, bufferManager buffer.BufferManager, parseResult *state.ParseResult, metatileSize int, tileSize int, metatileMaxDetailZoom int) (*state.MetatileResponseData, error) {
```

Comment:
this is all just pulling lines from the previous impl right?

Reply:
Yep. Pulling lines and adjusting variable names/return value.